Source/WebCore/html/ValidatedFormListedElement.h

```diff
@@ -0,0 +1,166 @@
+
+    bool isEnumeratable() const override { return false; }
+
+    bool wasInteractedWithSinceLastFormSubmitEvent() const { return m_wasInteractedWithSinceLastFormSubmitEvent; }
+    void setInteractedWithSinceLastFormSubmitEvent(bool);
+
+    bool matchesValidPseudoClass() const;
+    bool matchesInvalidPseudoClass() const;
+    bool matchesUserInvalidPseudoClass() const;
+    bool matchesUserValidPseudoClass() const;
+
+    virtual HTMLElement* anchorElement() = 0;
+
+    bool isCandidateForSavingAndRestoringState() const;
+    virtual bool shouldAutocomplete() const;
+    virtual bool shouldSaveAndRestoreFormControlState() const { return false; }
+    virtual FormControlState saveFormControlState() const;
+    virtual void restoreFormControlState(const FormControlState&) { } // Called only if state is not empty.
+    virtual const AtomString& formControlType() const = 0;
+
+protected:
+    virtual bool computeWillValidate() const;
+    void updateWillValidateAndValidity();
+    bool disabledByAncestorFieldset() const { return m_disabledByAncestorFieldset; }
+
+    bool validationMessageShadowTreeContains(const Node&) const;
+
+    virtual void disabledStateChanged();
+    virtual void disabledAttributeChanged();
+    virtual void readOnlyStateChanged();
+
+    void parseAttribute(const QualifiedName&, const AtomString&) override;
+    void insertedIntoAncestor(Node::InsertionType, ContainerNode&);
+    void removedFromAncestor(Node::RemovalType, ContainerNode&);
+    virtual void didFinishInsertingNode();
+    void finishParsingChildren();
+
+    void willChangeForm() override;
+    void didChangeForm() override;
+
+    void parseDisabledAttribute(const AtomString&);
+    void parseReadonlyAttribute(const AtomString&);
+
+    void resetDataListAncestorState();
+    void syncWithFieldsetAncestors(ContainerNode* insertionNode);
+
+private:
+    bool computeIsDisabledByFieldsetAncestor() const;
+
+    void startDelayingUpdateValidity() { ++m_delayedUpdateValidityCount; }
+    void endDelayingUpdateValidity();
+
+    std::unique_ptr<ValidationMessage> m_validationMessage;
+
+    // Cache of validity()->valid().
+    // But "candidate for constraint validation" doesn't affect isValid.
+    bool m_isValid : 1 { true };
+
+    // The initial value of willValidate depends on the derived class.
+    // We can't initialize it with a virtual function in the constructor.
+    // willValidate is not deterministic as long as willValidateInitialized is false.
+    mutable bool m_willValidate : 1 { true };
+    mutable bool m_willValidateInitialized : 1 { false };
+
+    bool m_disabled : 1 { false };
+    bool m_disabledByAncestorFieldset : 1 { false };
+
+    bool m_hasReadOnlyAttribute : 1 { false };
+    bool m_wasInteractedWithSinceLastFormSubmitEvent : 1 { false };
+    bool m_isFocusingWithValidationMessage { false };
+
+    enum DataListAncestorState { Unknown, InsideDataList, NotInsideDataList };
```

Comment:
Use `enum class DataListAncestorState : uint8_t`?

Reply:
Since we are slightly refactoring some old patterns anyway while moving the code, I went ahead and switched `m_dataListAncestorState` to `TriState m_isInsideDataList` in [3c92d2d](https://github.com/WebKit/WebKit/pull/4988/commits/3c92d2de75ee8a8c6cc58408b7734f25d2bc05d8). Seems a bit easier to grasp.